pkg/lightning/backend/local/duplicate.go

```diff
@@ -0,0 +1,637 @@
+	log.L().Info("Begin collect duplicate data from remote TiKV")
+	reqs, err := buildDuplicateRequests(tbl.Meta())
+	if err != nil {
+		return err
+	}
+
+	decoder, err := backendkv.NewTableKVDecoder(tbl, &backendkv.SessionOptions{
+		SQLMode: mysql.ModeStrictAllTables,
+	})
+	if err != nil {
+		return err
+	}
+	rpcctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	g, rpcctx := errgroup.WithContext(ctx)
+	for _, r := range reqs {
+		req := r
+		g.Go(func() error {
+			err := manager.sendRequestToTiKV(rpcctx, decoder, req)
+			if err != nil {
+				log.L().Error("error occur when collect duplicate data from TiKV", zap.Error(err))
+			}
+			return err
+		})
+	}
+	err = g.Wait()
+	log.L().Info("End collect duplicate data from remote TiKV")
+	return err
+}
+
+func (manager *DuplicateManager) sendRequestToTiKV(ctx context.Context,
+	decoder *backendkv.TableKVDecoder,
+	req *DuplicateRequest) error {
+	startKey := codec.EncodeBytes([]byte{}, req.start)
+	endKey := codec.EncodeBytes([]byte{}, req.end)
+
+	regions, err := paginateScanRegion(ctx, manager.splitCli, startKey, endKey, 1024)
+	if err != nil {
+		return err
+	}
+	tryTimes := 0
+	indexHandles := make([][]byte, 0)
+	for {
+		if len(regions) == 0 {
+			break
+		}
+		if tryTimes > maxRetryTimes {
+			return errors.Errorf("retry time exceed limit")
+		}
+		unfinishedRegions := make([]*split.RegionInfo, 0)
+		waitingClients := make([]sst.ImportSST_DuplicateDetectClient, 0)
+		watingRegions := make([]*split.RegionInfo, 0)
+		for idx, region := range regions {
+			if len(waitingClients) > manager.regionConcurrency {
+				r := regions[idx:]
+				unfinishedRegions = append(unfinishedRegions, r...)
+				break
+			}
+			_, start, _ := codec.DecodeBytes(region.Region.StartKey, []byte{})
+			_, end, _ := codec.DecodeBytes(region.Region.EndKey, []byte{})
+			if bytes.Compare(startKey, region.Region.StartKey) > 0 {
+				start = req.start
+			}
+			if region.Region.EndKey == nil || len(region.Region.EndKey) == 0 || bytes.Compare(endKey, region.Region.EndKey) < 0 {
+				end = req.end
+			}
+
+			cli, err := manager.getDuplicateStream(ctx, region, start, end)
+			if err != nil {
+				r, err := manager.splitCli.GetRegionByID(ctx, region.Region.GetId())
+				if err != nil {
+					unfinishedRegions = append(unfinishedRegions, region)
+				} else {
+					unfinishedRegions = append(unfinishedRegions, r)
+				}
+			} else {
+				waitingClients = append(waitingClients, cli)
+				watingRegions = append(watingRegions, region)
+			}
+		}
+
+		if len(indexHandles) > 0 {
+			handles := manager.getValues(ctx, indexHandles)
+			if len(handles) > 0 {
+				indexHandles = handles
+			} else {
+				indexHandles = indexHandles[:0]
+			}
+		}
+
+		for idx, cli := range waitingClients {
+			region := watingRegions[idx]
+			for {
+				resp, reqErr := cli.Recv()
+				hasErr := false
+				if reqErr != nil {
+					if errors.Cause(reqErr) == io.EOF {
+						break
+					} else {
+						hasErr = true
+					}
```

Comment:
```suggestion
					}
					hasErr = true
```

the linter should have barked at this 🤔

Reply:
Done